apis/python/src/tiledbsoma/_dataframe.py

```diff
@@ -422,7 +422,7 @@ def write(
 
                     # get new enumeration values, maintain original ordering
                     update_vals = []
-                    for new_val in col.chunk(0).dictionary.tolist():
+                    for new_val in col.dictionary.tolist():
```

Comment:
A suggestion:   this loop is going to be slow where the enumeration gets large.  We commonly see enum's with thousands of values. For example, in a recent Census dataset, the number of unique values in our "string label" columns ranges from 2 to 5576, with an average of about 450.

Another way to implement would be a hash table or set, e.g., `pandas.unique`.   I benchmarked a couple of examples with a few thousand enums, and using `unique` is roughly 40X faster on a trivial case, with a variety of enum lengths:

Two implementations:  `unique1` is essentially the loop above.  `unique2` does the same thing using concat/unique.

```
In [26]: def unique1(existing_cats, new_cats):
    ...:     update_vals = []
    ...:     for new_val in new_cats:
    ...:         if new_val not in existing_cats:
    ...:             update_vals.append(new_val)
    ...:     return existing_cats + update_vals
    ...: 
    ...: def unique2(existing_cats, new_cats):
    ...:     return pd.unique(np.concatenate((existing_cats, new_cats)))
    ...: 

In [27]: %timeit unique1( np.arange(1000).tolist(), np.arange(2000,0,-1).tolist())
7.16 ms ± 69 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [28]: %timeit unique2( np.arange(1000).tolist(), np.arange(2000,0,-1).tolist())
175 µs ± 306 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [29]: %timeit unique1( np.arange(400).tolist(), np.arange(500,400,-1).tolist())
204 µs ± 6.1 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [30]: %timeit unique2( np.arange(400).tolist(), np.arange(500,400,-1).tolist())
42.9 µs ± 955 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [31]: %timeit unique1( np.arange(5000).tolist(), np.arange(5500,4000,-1).tolist())
32.8 ms ± 1.35 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [32]: %timeit unique2( np.arange(5000).tolist(), np.arange(5500,4000,-1).tolist())
396 µs ± 12.9 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [33]: %timeit unique1( np.arange(10).tolist(), np.arange(11,20).tolist())
2.46 µs ± 117 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [34]: %timeit unique2( np.arange(10).tolist(), np.arange(11,20).tolist())
14.9 µs ± 64.2 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

```

So my thought is that using `unique` could improve the worst case by a lot, and doesn't hurt the trivial case very much.

Reply:
Thanks for the tip about using Pandas unique. I'll make ther modification tomorrow. 

Originally I had used Python's built-in set and set operations but had issues with the enum values getting shuffled around.

Reply:
Yeah, the nice thing about pandas.unique is that it preserves order.  From their docs:

> Uniques are returned in order of appearance. This does NOT sort.

Reply:
I had to use a different approach to get the set difference, but it is still faster than before:

```
import pandas as pd
import numpy as np

def unique1(existing_cats, new_cats):
    update_vals = []
    for new_val in new_cats:
        if new_val not in existing_cats:
            update_vals.append(new_val)
    return existing_cats + update_vals
    
def unique2(existing_cats, new_cats):
    return pd.Series(np.concatenate([existing_cats, existing_cats, new_cats])).drop_duplicates(keep=False)
```

```
%timeit unique1( np.arange(1000).tolist(), np.arange(2000,0,-1).tolist())
%timeit unique2( np.arange(1000).tolist(), np.arange(2000,0,-1).tolist())
8.68 ms ± 216 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
292 µs ± 10.8 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

```
%timeit unique1( np.arange(5000).tolist(), np.arange(5500,4000,-1).tolist())
%timeit unique2( np.arange(5000).tolist(), np.arange(5500,4000,-1).tolist())
40 ms ± 1.35 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
654 µs ± 27.4 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```